File: java-code/src/Solns/best_time_to_buy_and_sell_stock_121.java
//https://leetcode.com/problems/best-time-to-buy-and-sell-stock/submissions/

//Time complexity: O(n)
//Runtime: 3 ms, faster than 18.33% of Java online submissions for Best Time to Buy and Sell Stock.
//Memory Usage: 105.6 MB, less than 6.60% of Java online submissions for Best Time to Buy and Sell Stock.

public class best_time_to_buy_and_sell_stock_121 {
	public static void main(String[] args) {
		int[] a = {7,1,5,3,6,4};
		int[] b = {3,2,6,5,0,3};
		int[] c = {1,2,4};
		
		System.out.println(maxProfit(a));
		System.out.println(maxProfit(b));
		System.out.println(maxProfit(c));
	}
	
    public static int maxProfit(int[] prices) {

		var maxProfit = function(prices) {
			let MP = 0;
			
			let left = prices[0];
			let right = prices[1];
			
			for(let i = 1; i < prices.length - 1; i++){
				if(prices[i] < left)    
					left = prices[i];
				
				if(prices[i+1] < right)
					right = prices[i + 1]; 
				
				MP = Math.max(MP, right - left);
				
			}
			return MP;
		};
}

}

File: python-code/best_time_to_buy_and_sell_stock_121.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if(len(prices) <= 1):
            return 0
        left = 0
        maxProfit = prices[1] - prices[left]

        for right in range(1, len(prices)):
            if(prices[right] - prices[left] < 0 and right - left >= 2 or right == len(prices) - 1):
                print('increment left')
                left += 1
            else:
                print('increment right')
                right += 1

            maxProfit = max(prices[right] - prices[left], maxProfit)
        print(maxProfit)
        return maxProfit


    prices = [7,6,4,3,1]
    maxProfit(0, prices)